src/jace/jax/stages.py

```diff
@@ -0,0 +1,325 @@
+# JaCe - JAX Just-In-Time compilation using DaCe (Data Centric Parallel Programming)
+#
+# Copyright (c) 2024, ETH Zurich
+# All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+"""Reimplementation of the `jax.stages` module.
+
+This module reimplements the public classes of that Jax module.
+However, they are a big different, because Jace uses DaCe as backend.
+
+As in Jax Jace has different stages, the terminology is taken from [Jax' AOT-Tutorial](https://jax.readthedocs.io/en/latest/aot.html).
+- Stage out:
+    In this phase we translate an executable python function into Jaxpr.
+- Lower:
+    This will transform the Jaxpr into an SDFG equivalent.
+    As a implementation note, currently this and the previous step are handled as a single step.
+- Compile:
+    This will turn the SDFG into an executable object, see `dace.codegen.CompiledSDFG`.
+- Execution:
+    This is the actual running of the computation.
+"""
+
+from __future__ import annotations
+
+import copy
+from collections.abc import Callable, Mapping, Sequence
+from typing import Any, Final, TypeAlias
+
+import dace
+import jax as jax_jax
+
+from jace import optimization, translator, util
+from jace.jax import translation_cache as tcache
+from jace.translator import post_translation as ptrans
+from jace.util import dace_helper as jdace
+
+
+class Stage:
+    """A distinct step in the compilation chain, see module description for more.
+
+    The concrete steps are implemented in:
+    - JaceWrapped
+    - JaceLowered
+    - JaceCompiled
+    """
+
+
+"""Map type to pass compiler options to `JaceLowered.compile()`.
+"""
+CompilerOptions: TypeAlias = dict[str, tuple[bool, str]]
+
+
+class JaceWrapped(Stage):
+    """A function ready to be specialized, lowered, and compiled.
+
+    This class represents the output of functions such as `jace.jit()`.
+    Calling it results in jit (just-in-time) lowering, compilation, and execution.
+    It can also be explicitly lowered prior to compilation, and the result compiled prior to execution.
+
+    You should not create `JaceWrapped` instances directly, instead you should use `jace.jit`.
+
+    Todo:
+        Handles pytrees.
+        Copy the `jax._src.pjit.make_jit()` functionality to remove `jax.make_jaxpr()`.
+    """
+
+    _fun: Callable
+    _sub_translators: Mapping[str, translator.PrimitiveTranslatorCallable]
```

Comment:
The only thing that `PrimitiveTranslator` adds is basically the `.primitive` property, that you do not use in any way inside the driver.
The only place where it is important is inside the `register_primitive_subtranslator()` function and nowhere else.
Essentially, it imposes the fewest restrictions.

I do not remember exactly, but I started this rewrite, because MyPy complained about something, that was related to that.